### PR TITLE
Implement stochastic disaggregation for items and fm files

### LINKED ISSUE #1123: Stochastic disaggregation 4 & 6 File preparation for disaggregated locations
## Issue Description

This task covers the expanded items file feature for disaggregated locations (proposal 4), as well as the handling of financial terms (proposal 6) for the 6 different cases of OED location inputs

Use NumberOfBuildings from location file to generate expanded items file
Use IsAggregate flag value from location file to generate fm files.

Each disaggregated location has the same areaperil / vulnerability attributes as the parent coverage.

A new field is needed in gul_summary_map and fm_summary_map to link disaggregated locations to original location (disagg_id)

TIV, deductibles and limits are split equally.

The definition of site for the application of site terms depends on the value of IsAggregate. 

where IsAggregate = 1, site is the disaggregated location
where IsAggregate = 0, site is the non-disaggregated location

### PR DESCRIPTION
**Pull request by mtazzari**
This PR Fix #1123 by implementing the disaggregation in the items and fm files.

I am testing the behaviour using OasisPiWind model, branch [`feature/disaggregation`](https://github.com/OasisLMF/OasisPiWind/tree/feature/disaggregation), with:
```
oasislmf model run -C oasislmf_epa.json
````
where [`oasislmf_epa.json`](https://github.com/OasisLMF/OasisPiWind/blob/feature/disaggregation/oasislmf_epa.json).
Beware: there is an incorrect `epa_oasislmf.json` in the repository.

Questions/TODOs: 
 - [x] should we store somewhere the `disagg_id` field? I'd say no because it's a repetition of `item_id`, but let's check.
 - [ ] should we delete `epa_oasislmf.json` in OasisPiWind?
 - [ ] we need to add unit tests to cover all desired combinations of `IsAggregate` and `NumberOfBuildings`.

### Stochastic Disaggregation for items and fm files
Use `NumberOfBuildings` from location file to generate expanded items file
Use `IsAggregate` flag value from location file to generate fm files.

Each disaggregated location has the same areaperil / vulnerability attributes as the parent coverage.

A new field is needed in gul_summary_map and fm_summary_map to link disaggregated locations to original location (disagg_id)

TIV, deductibles and limits are split equally.

The definition of site for the application of site terms depends on the value of IsAggregate.

where `IsAggregate` = 1, site is the disaggregated location
where `IsAggregate` = 0, site is the non-disaggregated location

### CODE REVIEW

File: oasislmf/preparation/gul_inputs.py [hunk 1]
```diff
@@ -169,10 +170,20 @@ def get_gul_input_items(
...
 
+    # add default values if missing
+    if 'IsAggregate' not in location_df.columns:
+        location_df['IsAggregate'] = 0
+        location_df['IsAggregate'].fillna(0, inplace=True)
+
+    # Make sure NumberOfBuildings is there and filled (not mandatory), otherwise assume NumberOfBuildings = 1
+    if 'NumberOfBuildings' not in location_df.columns:
+        location_df['NumberOfBuildings'] = 1
+        location_df['NumberOfBuildings'].fillna(1, inplace=True)
```
Comment (sstruzik) [inline oasislmf/preparation/gul_inputs.py:181]: both fillna added, Probably need to be put in the else

File: oasislmf/preparation/gul_inputs.py [hunk 2]
```diff
@@ -273,83 +283,94 @@ def get_gul_input_items(
...
-    gul_inputs_df = pd.concat(gul_inputs_reformatted_chunks)
-    gul_inputs_df[list(terms_found)].fillna(0., inplace=True)
-    gul_inputs_df = gul_inputs_df.sort_index().reset_index(drop=True)
-    term_int_found = list(set(gul_inputs_df.columns).intersection(set(term_cols_ints + terms_ints)))
-    gul_inputs_df[term_int_found] = gul_inputs_df[term_int_found].fillna(0)
-    # Set default values and data types for BI coverage boolean, TIV, deductibles and limit
+        terms_found.update(cols_by_cov_type[cov_type]['column_mapping_dict'].values())
+
+        # if NumberOfBuildings == 0: still add one entry
+        disagg_df = pd.concat([cov_type_group] * max(number_of_buildings, 1), ignore_index=True)
```
Comment (sstruzik) [inline oasislmf/preparation/gul_inputs.py:346]: I think if you ignore index here you cannot sort index later on in:
    gul_inputs_df = (
        pd.concat(gul_inputs_reformatted_chunks)
        .fillna(value={c: 0 for c in terms_found})
        .sort_index()
        .reset_index(drop=True)
        .fillna(value={c: 0 for c in set(gul_inputs_df.columns).intersection(set(term_cols_ints + terms_ints))})
    )

File: oasislmf/preparation/gul_inputs.py [hunk 3]
```diff
@@ -376,24 +397,26 @@ def get_gul_input_items(
...
-        ['is_bi_coverage', 'group_id', 'coverage_id', 'item_id', 'status']
+        ['is_bi_coverage', 'group_id', 'coverage_id', 'item_id', 'status', 'disagg_id', 'NumberOfBuildings', 'IsAggregate',] +  # disagg_id is needed for fm_summary_map
+        (["peril_correlation_group", "damage_correlation_value", 'hazard_group_id', "hazard_correlation_value"] if correlations is True else [])
     )
-    if correlations is True:
-        usecols += ["peril_correlation_group", "damage_correlation_value", 'hazard_group_id', "hazard_correlation_value"]
 
-    usecols = [col for col in usecols if col in gul_inputs_df]
-    gul_inputs_df = gul_inputs_df[usecols]
+    usecols = list(set(usecols) & set(gul_inputs_df.columns))
```
Comment (sstruzik) [inline oasislmf/preparation/gul_inputs.py:412]: I think this give a different order for the columns and make the header of gul_summary different

**Review by johcarter — APPROVED** [review]
I've tested all combinations of IsAggregate and NumberOfBuildings on PiWind and it looks good for release.  I'll work on adding tests to tests/fm for the subsequent release.

**Review by sambles — APPROVED** [review]

**Pull request merged**